### ISSUE TITLE
urweb road map

### ISSUE THREAD

**Issue by ansarizafar** [urweb/urweb#99]
UrWeb is an amazing project and It can disrupt web development. Is there any road map available? Is there any plane to update urweb and use modern technologies like web sockets, web works, service works etc.

**Comment by achlipala** [comment]
I'm glad to hear that you're so excited about the Ur/Web concept!  I am not familiar in detail with any of the ingredients you mentioned.  Could you summarize how the programmer or user experience would benefit from adopting them?  (There's a good chance that most of them could be incorporated with zero changes at the source-code level, just requiring changes to the Ur/Web compiler or runtime system, but I'm not clear on what the payoff would be.)

**Comment by ansarizafar** [comment]
Web development is very difficult. I have tried Asp.net, Nodejs, Elixir, Crystal Language, Pony lang, Nim lang for writing server side logic and Angular, React, Aurelia, Vuejs, Polymer and Stenciljs for writing client side app. One has to learn so many things (Serer side language, Javaacript, Html, Css, Webpack configurations etc) to develop a modern app. UrWeb is the best solution, I have seen so far but I am surprised, Why UrWeb is not popular and actively developed. Progressive web apps are a buzz word and the future of web development. All modern browsers now support modern features like ES6/7, Web components, Web workers, Service workers etc and these features are necessary for building modern progressive web apps. Please visit these links.

https://developers.google.com/web/progressive-web-apps/
https://developers.google.com/web/fundamentals/primers/service-workers/
https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers
https://developer.mozilla.org/en-US/docs/Web/Web_Components
https://stenciljs.com/
 
Web developers all around the world are in search of a better full stack solution like urweb. I firmly believe that by adding modern browser features, Better UrWeb website, documentations/tutorials for the beginners, better marketing and community building effects we can make Urweb a de facto standard for web development. The time is right for UrWeb to come out from academic circles and dominate the world of web development.

**Comment by achlipala** [comment]
Let me clarify that Ur/Web _is_ actively developed.  It's just that folks mostly aren't asking for small-to-medium-sized features, and big features will have to wait for a future redesign!  I still put out a few new releases each year, which go into Debian, Homebrew, etc.

Would you mind sketching a concrete scenario or two, showing benefits of adding some of those Web standards to Ur/Web?  You've pointed to more text than I have time to read just now.

Also, maybe it's worth taking this discussion to the ur mailing list, where more people can participate.

**Comment by ansarizafar** [comment]
All the above mentioned modern browser features are must for building progressive web apps with UrWeb.

- Progressive web apps  -- A new way to deliver amazing user experiences on the web.

- Es6/7 -- ES6/7 is a significant update to the javascript language, and the first update to the language since ES5 was standardized in 2009.

- Service wrokers -- provides rich offline experiences, periodic background syncs, push notifications—functionality that would normally require a native application—are coming to the web. Service workers provide the technical foundation that all these features rely on.

- Web workers -- Web Workers is a simple means for web content to run scripts in background threads. The worker thread can perform tasks without interfering with the user interface. In addition, they can perform I/O using XMLHttpRequest.

- Web Components -- Web Components consists of several separate technologies. You can think of Web Components as reusable user interface widgets that are created using open Web technology. They are part of the browser, and so they do not need external libraries like jQuery or Dojo. An existing Web Component can be used without writing code, simply by adding an import statement to an HTML page. Web Components use new or still-developing standard browser capabilities.

I know your are very busy, but I would request you to visit these links whenever you can spare some time.

I would suggest Github issues for discussion as all web developers mostly use Github. I wouls also suggest http://gitter.im/

**Comment by ratmice** [comment]
Another interesting thing for the roadmap would be a webassembly backend,
Ur/Web really seems to follow the webassembly model closely with few(any?) corner cases with regards to the confinement model and the current lack of GC in wasm doesn't seem to be an issue.

I'm not sure exactly what that means though,
the obvious thing would be to emit client-side code as wasm, 

however we could also emit server side code as wasm with emscripten having run sqlite,
with that I suppose you could run the demos without having to host a database.

I suppose the path of least resistance currently would be just running the c code output by the current compiler through emscripten as well

**Comment by achlipala** [comment]
I bet it's not a huge job to target WebAssembly, and it might be worthwhile, though I remain skeptical that it actually helps anyone very much.

**Comment by ansarizafar** [comment]
WebAssembly support now shipping in all major browsers https://blog.mozilla.org/blog/2017/11/13/webassembly-in-browsers/

**Comment by ansarizafar** [comment]
Great opportunity for urweb [Are You Ready for the WebAssembly Revolution?](https://thenewstack.io/ready-web-assembly-revolution/?utm_content=bufferb42a4&utm_medium=social&utm_source=twitter.com&utm_campaign=buffer)
